src/protobuf-net/Meta/RuntimeTypeModel.cs

```diff
@@ -719,15 +719,34 @@ internal int FindOrAddAuto(Type type, bool demand, bool addWithContractOnly, boo
                             shouldAdd = addEvenIfAutoDisabled = true; // always add basic tuples, such as KeyValuePair
                         }
 
-                        if (!shouldAdd || (
-                            !type.IsEnum && addWithContractOnly && family == MetaType.AttributeFamily.None)
-                            )
+                        // for generic types, when no contract is set, see if a generic definition is registered and use that
+                        if (type.IsGenericType && !type.IsGenericTypeDefinition && family == MetaType.AttributeFamily.None)
                         {
-                            if (demand) ThrowUnexpectedType(type, this);
-                            return key;
+                            var genericTypeDefinition = type.GetGenericTypeDefinition();
+                            key = types.IndexOf(MetaTypeFinder, genericTypeDefinition);
+                            if (key >= 0)
+                            {
+                                var genericMetaType = (MetaType)types[key];
+                                if (genericMetaType.Pending)
+                                {
+                                    WaitOnLock();
+                                }
+                                metaType = genericMetaType.GenerateGenericType(type);
+                            }
                         }
 
-                        metaType = Create(type);
+                        if (metaType is null)
+                        {

```

Comment:
This additional check seemed like the best solution to avoid triggering the next block if a metatype was generated in the new code I wrote. In my head it was either this or a `goto` statement.